### ISSUE TITLE
QuantumRegister indexing is recursive and shouldn't be

### ISSUE THREAD

**Issue by Woody1193** [Qiskit/qiskit-terra#1964]
### Information

- **Qiskit Terra version**: 0.7.1
- **Python version**: 3.5.6
- **Operating system**: Windows 7

### Steps to reproduce the problem

```
q = QuantumRegister(1)
def inf_recurse(qarg):
    print( qarg[0] )
    inf_recurse( qarg[0] )
```

### What is the current behavior?

When indexing a QuantumRegister, indexing can be done recursively and therefore infinitely.

```
> QuantumRegister(1, 'q0')
> (QuantumRegister(1, 'q0'), 0)
> QuantumRegister(1, 'q0')
> (QuantumRegister(1, 'q0'), 0)
> QuantumRegister(1, 'q0')
> (QuantumRegister(1, 'q0'), 0)
> QuantumRegister(1, 'q0')
> (QuantumRegister(1, 'q0'), 0)
> QuantumRegister(1, 'q0')
> (QuantumRegister(1, 'q0'), 0)
> File "qiskit-terra\qiskit\circuit\register.py", line 77, in __getitem__
     self.check_range(key)
   File "qiskit-terra\qiskit\circuit\register.py", line 59, in check_range
     if j < 0 or j >= self.size:
 RecursionError: maximum recursion depth exceeded in comparison
```

### What is the expected behavior?

```
> QuantumRegister(1, 'q0')
> Qubit('q0_0')
> TypeError. Qubit object is not subscriptable
```

### Suggested solutions

My first suggestion would be to restructure the `QuantumRegister` as a list of some inner Qubit object. Since these objects are only used for book-keeping on the qubits, it would make more sense to separate the collection functionality and the qubit functionality into two classes.

**Comment by kdk** [comment]
See also #1647, which raises a similar concern with the usability of `QuantumRegister`.

**Comment by Woody1193** [comment]
@kdk The fixes that would address this issue would also address my issue. Should I close/delete this then? They're not exactly the same but the fix for #1647 would also fix mine.

**Comment by kdk** [comment]
@Woody1193 Great, I'll close this then.
